server/src/internalClusterTest/java/org/opensearch/snapshots/SearchableSnapshotIT.java

```diff
@@ -60,109 +60,191 @@ protected Settings.Builder randomRepositorySettings() {
-        indexRandomDocs("test-idx-1", 100);
-        indexRandomDocs("test-idx-2", 100);
 
-        logger.info("--> snapshot");
-        final CreateSnapshotResponse createSnapshotResponse = client.admin()
-            .cluster()
-            .prepareCreateSnapshot("test-repo", "test-snap")
-            .setWaitForCompletion(true)
-            .setIndices("test-idx-1", "test-idx-2")
-            .get();
-        MatcherAssert.assertThat(createSnapshotResponse.getSnapshotInfo().successfulShards(), greaterThan(0));
-        MatcherAssert.assertThat(
-            createSnapshotResponse.getSnapshotInfo().successfulShards(),
-            equalTo(createSnapshotResponse.getSnapshotInfo().totalShards())
-        );
+        internalCluster().ensureAtLeastNumDataNodes(Math.max(numReplicasIndex1, numReplicasIndex2) + 1);
+        createIndexWithDocsAndEnsureGreen(numReplicasIndex1, "test-idx-1");
+        createIndexWithDocsAndEnsureGreen(numReplicasIndex2, "test-idx-2");
 
-        assertTrue(client.admin().indices().prepareDelete("test-idx-1", "test-idx-2").get().isAcknowledged());
+        takeSnapshot(client, "test-idx-1", "test-idx-2");
+        deleteIndicesAndEnsureGreen(client, "test-idx-1", "test-idx-2");
 
         internalCluster().ensureAtLeastNumSearchNodes(Math.max(numReplicasIndex1, numReplicasIndex2) + 1);
+        restoreSnapshotAndEnsureGreen(client);
 
-        logger.info("--> restore indices as 'remote_snapshot'");
+        assertDocCount("test-idx-1-copy", 100L);
+        assertDocCount("test-idx-2-copy", 100L);
+        assertIndexDirectoryDoesNotExist("test-idx-1-copy", "test-idx-2-copy");
+    }
+
+    /**
+     * Tests the functionality of remote shard allocation to
+     * ensure it can assign remote shards to a node with local shards given it has the
+     * {@link org.opensearch.cluster.node.DiscoveryNodeRole#SEARCH_ROLE} capabilities.
+     */
+    public void testSearchableSnapshotAllocationForLocalAndRemoteShardsOnSameNode() throws Exception {
+        final int numReplicasIndex = randomIntBetween(1, 4);
+        final String indexName = "test-idx";
+        final String restoredIndexName = indexName + "-copy";
+        final Client client = client();
+
+        internalCluster().ensureAtLeastNumSearchAndDataNodes(numReplicasIndex + 1);
+        createIndexWithDocsAndEnsureGreen(numReplicasIndex, indexName);
+        takeSnapshot(client, indexName);
+
+        restoreSnapshotAndEnsureGreen(client);
+
+        assertDocCount(restoredIndexName, 100L);
+        assertDocCount(indexName, 100L);
+    }
+
+    /**
+     * Tests the functionality of remote shard allocation to
+     * ensure it can handle node drops for failover scenarios and the cluster gets back to a healthy state when
+     * nodes with search capabilities are added back to the cluster.
+     */
+    public void testSearchableSnapshotAllocationForFailoverAndRecovery() throws Exception {
+        final int numReplicasIndex = randomIntBetween(1, 4);
```

Comment:
What do you think about taking the randomness away here and setting number of replicas to 1. Then you could run the following scenario:

1. Ensure green after initial setup
2. Stop one search node and ensure yellow
3. Stop another search node and ensure red
4. Add 3 new search nodes and ensure green
5. Stop one search node and ensure green

Then I think you could get rid of the "testSearchableSnapshotAllocationWithNoSearchNodes" test case since that is covered here. It would also cover the "automatically recovery from red" scenario that currently isn't being tested.

Reply:
This led to a discovery of an issue when the primary remote shards go unassigned. I'll add in the fix for it as well. 
Thanks for the suggestion 🙂